File: java/trip-service-kata/src/test/java/org/craftedsw/tripservicekata/trip/TripServiceTest.java
package org.craftedsw.tripservicekata.trip;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import org.craftedsw.tripservicekata.exception.UserNotLoggedInException;
import org.craftedsw.tripservicekata.user.Guest;
import org.craftedsw.tripservicekata.user.User;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

public class TripServiceTest {

  private static final Trip MADRID = new Trip();
  private static final Trip BARCELONA = new Trip();
  private static final Trip BUENOS_AIRES = new Trip();

  TripService tripService;

  User loggedUser;
  User anotherUser;

  List<Trip> anotherUserTrips = new ArrayList<>();

  List<Trip> returnedTrips = new ArrayList<>();

  @BeforeEach
  void beforeEach(){
    this.tripService = new TestableTripService();
  }

  @Test
  void forNotLoggedUserThrowException(){
    givenGuestIsLogged();
    givenAnotherUserWithNoFriends();

    assertThrows(
        UserNotLoggedInException.class,
        this::whenGettingTripsByUser
    );
  }

  @Test
  void forLoggedUserNotAFriendWithOtherCantSeeTrips(){
    givenLoggedUserWithNoFriends();
    givenAnotherUserWithNoFriends();

    whenGettingTripsByUser();

    thenShouldGetNumberOfTips(0);
  }

  @Test
  void forLoggedUserFriendedWithOtherCanSeeTrips(){
    givenLoggedUserWithNoFriends();
    givenAnotherUserWithLoggedFriend();
    givenAnotherUserHasTripsTo(MADRID, BARCELONA, BUENOS_AIRES);

    whenGettingTripsByUser();

    thenShouldGetNumberOfTips(2);
  }

  private void givenAnotherUserWithNoFriends() {
    this.anotherUser = new User();
  }

  private void givenAnotherUserWithLoggedFriend() {
    this.anotherUser = new User();
    this.anotherUser.addFriend(loggedUser);
  }

  private void givenLoggedUserWithNoFriends() {
    loggedUser = new User();
  }

  private void givenGuestIsLogged() {
    loggedUser = new Guest();
  }

  private void givenAnotherUserHasTripsTo(Trip... trips) {
    anotherUserTrips.addAll(Arrays.asList(trips));
    anotherUserTrips.forEach(trip -> anotherUser.addTrip(trip));
  }

  private void whenGettingTripsByUser() {
    returnedTrips = this.tripService.getTripsByUser(anotherUser);
  }

  private void thenShouldGetNumberOfTips(int numberOfTrips) {
    assertEquals(
        numberOfTrips, returnedTrips.size(),
        "Number of trips should be " + numberOfTrips 
    );
  } 


  private class TestableTripService extends TripService {

    @Override
    protected User loggedUser() {
      return loggedUser;
    }

  }
}

File: kotlin/trip-service-kata/src/test/kotlin/org/craftedsw/tripservicekata/trip/TripServiceTest.kt
package org.craftedsw.tripservicekata.trip

import org.craftedsw.tripservicekata.exception.UserNotLoggedInException
import org.craftedsw.tripservicekata.user.User
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import java.util.*

class TripServiceTest {

    val ANY_USER = User()
    val LONDON = Trip()

    private lateinit var tripService: TripServiceTestable

    private var user: User = User()
    private var loggedUser: User = User()

    private var usersTripList: MutableList<Trip> = mutableListOf()
    private var tripList: List<Trip> = emptyList()

    @Test
    fun `should throw user not logged exception when user is a guest`(){
        givenGuestUser()

        assertThrows<UserNotLoggedInException> {
            whenGetTripsByUserIsCalled()
        }
    }

    @Test
    fun `should return empty trip list when user has no friends`(){
        givenLoggedUser()
        givenUserHasNoFriends()

        whenGetTripsByUserIsCalled()

        thenTripListShouldBeEmpty()
    }

    @Test
    fun `should return empty trip list when user is not friends with logged user`(){
        givenLoggedUser()
        givenUserIsFriendsWith(ANY_USER)

        whenGetTripsByUserIsCalled()

        thenTripListShouldBeEmpty()
    }

    @Test
    fun `should return user's trip list when user is friends with logged user`(){
        givenLoggedUser()
        givenUserIsFriendsWith(loggedUser)
        givenUserHasATripTo(LONDON)

        whenGetTripsByUserIsCalled()

        thenTripListShouldHaveTripsTo(listOf(LONDON))
    }

    private fun givenGuestUser(){
        tripService = TripServiceTestable(null)
    }

    private fun givenLoggedUser(){
        loggedUser = User()
        tripService = TripServiceTestable(loggedUser)
    }

    private fun givenUserHasNoFriends() {
        user = User()
    }

    private fun givenUserIsFriendsWith(anyUser: User) {
        user.addFriend(anyUser)
    }

    private fun givenUserHasATripTo(trip: Trip) {
        usersTripList.add(trip)
        tripService = TripServiceTestable(loggedUser, usersTripList)
    }

    private fun whenGetTripsByUserIsCalled() {
        tripList = tripService.getTripsByUser(user)
    }

    private fun thenTripListShouldBeEmpty() {
        assertEquals(Collections.emptyList<Trip>(), tripList)
    }

    private fun thenTripListShouldHaveTripsTo(trips: List<Trip>) {
        assertEquals(trips, tripList)
    }

    class TripServiceTestable(
            private var loggedUser: User? = User(),
            private val trips: List<Trip> = ArrayList()) : TripService(){

        override fun getLoggedUser() = loggedUser
        override fun findTripsFor(user: User) = trips
    }

}
